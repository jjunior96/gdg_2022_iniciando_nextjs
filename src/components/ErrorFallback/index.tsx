import { useRouter } from 'next/router';

import Button from 'components/Button';

import * as S from './styles';

export type ErrorFallbackProps = {
  error: {
    message: string;
  };
  resetErrorBoundary?: () => void;
};

const ErrorFallback = ({ error, resetErrorBoundary }: ErrorFallbackProps) => {
  const route = useRouter();

  const handlePushToHome = () => {
    route.push('/');
  };

  return (
    <S.Container role="alert">
      <S.Content>
        <S.Title>Something went wrong</S.Title>
        <S.Description>{error.message}</S.Description>

        <Button onClick={resetErrorBoundary || handlePushToHome}>
          Try again
        </Button>
      </S.Content>
    </S.Container>
  );
};

export default ErrorFallback;
