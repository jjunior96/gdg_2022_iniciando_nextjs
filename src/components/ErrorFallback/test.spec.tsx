/* eslint-disable @typescript-eslint/no-explicit-any */
import { useRouter } from 'next/router';

import { fireEvent, screen, waitFor } from '@testing-library/react';

import { renderWithTheme } from 'utils/tests/__config__/helpers';

import ErrorFallback from '.';

jest.mock('next/router', () => ({
  __esModule: true,
  useRouter: jest.fn()
}));

const handlePushToHome = jest.fn();

const errorMock = {
  message: 'something wrong'
};

describe('<ErrorFallback />', () => {
  it('should render component by default', () => {
    const { container } = renderWithTheme(<ErrorFallback error={errorMock} />);

    expect(
      screen.getByRole('heading', { name: /something went wrong/i })
    ).toBeInTheDocument();

    expect(screen.getByText(errorMock.message)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should to run function when passed', async () => {
    renderWithTheme(
      <ErrorFallback error={errorMock} resetErrorBoundary={handlePushToHome} />
    );

    fireEvent.click(screen.getByRole('button', { name: /try again/i }));

    await waitFor(() => {
      expect(handlePushToHome).toHaveBeenCalledTimes(1);
    });
  });

  it('should to run function internal when not passed other function', async () => {
    const mockRouter = {
      push: jest.fn()
    };
    (useRouter as jest.Mock).mockReturnValue(mockRouter);

    renderWithTheme(<ErrorFallback error={errorMock} />);

    fireEvent.click(screen.getByRole('button', { name: /try again/i }));

    await waitFor(() => {
      expect(handlePushToHome).toHaveBeenCalledTimes(1);
    });
  });
});
