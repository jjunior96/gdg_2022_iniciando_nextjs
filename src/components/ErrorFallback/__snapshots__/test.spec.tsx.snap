// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorFallback /> should render component by default 1`] = `
.c4 {
  width: 100%;
  color: #FAFAFA;
  background-color: #24A66B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.4rem;
  border: 0;
  padding: 0 1.6rem;
  font-weight: 600;
  font-size: 1.6rem;
  -webkit-transition: background-color 0.3s,box-shadow 0.2s;
  transition: background-color 0.3s,box-shadow 0.2s;
  outline: none;
  height: 5.6rem;
  font-size: 1.6rem;
}

.c4:focus {
  outline: 3px solid #aaecce;
}

.c4:hover {
  background-color: #29bb79;
  box-shadow: 0 8px 8px rgba(0,0,0,0.2);
}

.c4 > svg {
  margin-right: 1rem;
}

.c0 {
  width: 100%;
}

.c1 {
  width: 100%;
  max-width: 46rem;
  margin: 0 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  font-size: 3.2rem;
  font-weight: 600;
  margin-bottom: 1.6rem;
}

.c3 {
  font-size: 2.4rem;
  font-weight: 400;
  text-align: center;
  margin-bottom: 2.4rem;
}

<div
  class="c0"
  role="alert"
>
  <div
    class="c1"
  >
    <h2
      class="c2"
    >
      Something went wrong
    </h2>
    <p
      class="c3"
    >
      something wrong
    </p>
    <button
      class="c4"
      color="green"
    >
      Try again
    </button>
  </div>
</div>
`;
